src/waveform/renderers/allshader/waveformrendererstem.cpp

```diff
@@ -0,0 +1,202 @@
+    TrackPointer pTrack = m_waveformRenderer->getTrackInfo();
+    if (!pTrack || (m_isSlipRenderer && !m_waveformRenderer->isSlipActive())) {
+        return;
+    }
+
+    auto stemInfo = pTrack->getStemInfo();
+    // If this track isn't a stem track, skip the rendering
+    if (stemInfo.isEmpty()) {
+        return;
+    }
+    auto positionType = m_isSlipRenderer ? ::WaveformRendererAbstract::Slip
+                                         : ::WaveformRendererAbstract::Play;
+
+    ConstWaveformPointer waveform = pTrack->getWaveform();
+    if (waveform.isNull()) {
+        return;
+    }
+
+    const int dataSize = waveform->getDataSize();
+    if (dataSize <= 1) {
+        return;
+    }
+
+    const WaveformData* data = waveform->data();
+    if (data == nullptr) {
+        return;
+    }
+    // If this waveform doesn't contain stem data, skip the rendering
+    if (!waveform->hasStem()) {
+        return;
+    }
+
+    const float devicePixelRatio = m_waveformRenderer->getDevicePixelRatio();
+    const int length = static_cast<int>(m_waveformRenderer->getLength() * devicePixelRatio);
+
+    // See waveformrenderersimple.cpp for a detailed explanation of the frame and index calculation
+    const int visualFramesSize = dataSize / 2;
+    const double firstVisualFrame =
+            m_waveformRenderer->getFirstDisplayedPosition(positionType) * visualFramesSize;
+    const double lastVisualFrame =
+            m_waveformRenderer->getLastDisplayedPosition(positionType) * visualFramesSize;
+
+    // Represents the # of visual frames per horizontal pixel.
+    const double visualIncrementPerPixel =
+            (lastVisualFrame - firstVisualFrame) / static_cast<double>(length);
+
+    // Per-band gain from the EQ knobs.
+    float allGain(1.0);
+    // applyCompensation = true, as we scale to match filtered.all
+    getGains(&allGain, false, nullptr, nullptr, nullptr);
+
+    const float breadth = static_cast<float>(m_waveformRenderer->getBreadth()) * devicePixelRatio;
+    const float halfBreadth = breadth / 2.0f;
+
+    const float heightFactor = allGain * halfBreadth / m_maxValue;
+
+    // Effective visual frame for x
+    double xVisualFrame = qRound(firstVisualFrame / visualIncrementPerPixel) *
+            visualIncrementPerPixel;
+
+    const int numVerticesPerLine = 6; // 2 triangles
+
+    const int reserved = numVerticesPerLine * (8 * length + 1);
+
+    m_vertices.clear();
+    m_vertices.reserve(reserved);
+    m_colors.clear();
+    m_colors.reserve(reserved);
+
+    m_vertices.addRectangle(0.f,
+            halfBreadth - 0.5f * devicePixelRatio,
+            static_cast<float>(length),
+            m_isSlipRenderer ? halfBreadth : halfBreadth + 0.5f * devicePixelRatio);
+    m_colors.addForRectangle(0.f, 0.f, 0.f, 0.f);
+
+    const double maxSamplingRange = visualIncrementPerPixel / 2.0;
+
+    for (int pos = 0; pos < length; ++pos) {
+        for (int s = 0; s < 4; s++) {
+            for (int z = 0; z < 2; z++) {
```

Comment:
```suggestion
    for (int pixelPos = 0; pixelPos < length; ++pixelPos) {
        for (int stemIdx = 0; stemIdx < kMaxSupportedStem; stemIdx++) {
            for (int chIdx = 0; chIdx < mixxx::audio::ChannelCount::stereo(); chIdx++) {
```
Please use self descriptive names for the indices - I'm still unsure if I got the mening right.

Reply:
`z` stands for `z-index` - signal is drawn twice to help visualising the maximum signal when the stem volume isn't 100% or muted.
I added a comment.